### ISSUE TITLE
Intersection of discriminated union (of records) and a record is incorrect

### ISSUE THREAD

**Issue by RebeccaStevens** [microsoft/TypeScript#47612]
# Bug Report

### 🔎 Search Terms

intersection with discriminated union

This might be a duplicate of #33654 but I'm not completely sure. This issue has a simpler example anyway.
It also seems to be a duplicate of #9919 which was reportedly fixed. This is either regression or a missed case.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (v3.3 and v4.5), and I reviewed the FAQ for entries about intersection with discriminated union.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAglC8UDeUCGAuKBnYAnAlgHYDmA3FAEaaECuAthRLlAL5QBkyamNhA1oQD2Ad0KtSAKFCQoAIQSwA2gCJUygLrkA9Fux4ixCVPDQAGgoAUKDHoIlWUAD5cqUWgyasAlBy43eAiJiLJLS0ACaCqYqappQOlABQqISQA)

### 💻 Code

```ts
type X = ({ a: string } | { b: number }) & { a: unknown };
type Y = X["a"]; // actual: unknown, expected: string
```

### 🙁 Actual behavior

type `Y` is `unknown`.

### 🙂 Expected behavior

type `Y` should be `string`.

Why:

  - When discriminating `{ a: string } | { b: number }` with "`a` must be defined as a field", we get `{ a: string }`.
  - `{ a: string } & { a: unknown }` can be simplified to `{ a: string }`

**Comment by RyanCavanaugh** [comment]
This is the intended behavior. Through distributivity:
```
type X = ({ a: string } | { b: number }) & { a: unknown };
// is equivalent to
type X = ({ a: string } & { a: unknown }) | ({ b: number } & { a: unknown });
//                                          ^^^^^^^^^ this one  ^^^^^^^^^^^^```
```
so `{ b: 3, a: "foo" }` is a legal inhabitant of the second constituent of the union, so given an arbitrary value of `x` you can't be guaranteed to get a `string` from it

**Comment by RebeccaStevens** [comment]
That makes sense.

What's different about the follow case then? How come TypeScript is able to determine that the resulting type is `string` here?

```ts
declare const foo: { a: string } | { b: number };
declare function hasA(value: unknown): value is typeof value & { a: unknown };

if (hasA(foo)) {
    type Z = typeof foo['a']; // string
}
```

[Playground](https://www.typescriptlang.org/play?ts=4.5.4#code/CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAN7xSGYwCWKA5vAL7wA+x8ARoSgK4C2bIMBgG4AUKEiwE2LijAZKqeAAsoaAIIAKAG5QIXEIRkBrFEgDuKAJSEdehJTTwMATwAOIJNni398AGSsZPDGphbCIiKUXhoq6hq4SJaWxCLwaU5uCABa8AC8Ge6eOHgA2gDkUGUAukLwAPR18BTUNCL0QA)
